src/compute/src/sink/persist_sink.rs

```diff
@@ -101,65 +102,60 @@ where
             scope,
             move |_capabilities, frontiers, scheduler| async move {
                 let mut buffer = Vec::new();
-                let mut stash: HashMap<Timestamp, Vec<((Row, Row), Timestamp, Diff)>> =
-                    HashMap::new();
-
-                // Keep an empty Row that we can re-use when we are replacing an Option with an empty
-                // Row. Our output type is `(Row, Row)` instead of `(Option<Row>, Option<Row>)`.
-                let empty_row = Row::default();
+                let mut stash: HashMap<_, Vec<_>> = HashMap::new();
 
                 // TODO(aljoscha): We need to figure out what to do with error results from these calls.
-                let persist_client = persist_location
+                let mut write = persist_location
                     .open()
                     .await
-                    .expect("cannot open persist client");
-
-                let mut write = persist_client
+                    .expect("could not open persist client")
                     .open_writer::<Row, Row, Timestamp, Diff>(shard_id)
                     .await
                     .expect("could not open persist shard");
 
                 while scheduler.notified().await {
-                    let frontier = frontiers.borrow()[0].clone();
+                    let input_frontier = frontiers.borrow()[0].clone();
 
-                    if !active_write_worker || token_weak.upgrade().is_none() {
+                    if !active_write_worker
+                        || token_weak.upgrade().is_none()
+                        || shared_frontier.borrow().is_empty()
+                    {
                         return;
                     }
 
                     input.for_each(|_cap, data| {
                         data.swap(&mut buffer);
 
-                        let mapped_updates = buffer.drain(..).map(|((key, value), ts, diff)| {
-                            let key = key.unwrap_or_else(|| empty_row.clone());
-                            let value = value.unwrap_or_else(|| empty_row.clone());
-                            ((key, value), ts, diff)
-                        });
-
-                        for update in mapped_updates {
-                            stash.entry(update.1).or_default().push(update);
+                        for ((key, value), ts, diff) in buffer.drain(..) {
+                            let key = key.unwrap_or_default();
+                            let value = value.unwrap_or_default();
```

Comment:
I assume creating a default row is no more expensive than cloning an existing one, so there is no reason to not use this simpler code. Let me know if I missed something!